contracts/utils/Strings.sol

```diff
@@ -115,4 +128,247 @@ library Strings {
+        return tryParseUint(input, 0, bytes(input).length);
+    }
+
+    /**
+     * @dev Variant of {parseUint-string-uint256-uint256} that returns false if the parsing fails because of an invalid
+     * character.
+     *
+     * This function will still revert if the result does not fit in a `uint256`
+     */
+    function tryParseUint(
+        string memory input,
+        uint256 begin,
+        uint256 end
+    ) internal pure returns (bool success, uint256 value) {
+        bytes memory buffer = bytes(input);
+
+        uint256 result = 0;
+        for (uint256 i = begin; i < end; ++i) {
+            uint8 chr = _tryParseChr(buffer[i]);
+            if (chr > 9) return (false, 0);
+            result *= 10;
+            result += chr;
+        }
+        return (true, result);
+    }
+
+    /**
+     * @dev Parse a decimal string and returns the value as a `int256`.
+     *
+     * This function will revert if:
+     * - the string contains any character (outside the prefix) that is not in [0-9].
+     * - the result does not fit in a `int256`.
+     */
+    function parseInt(string memory input) internal pure returns (int256) {
+        return parseInt(input, 0, bytes(input).length);
+    }
+
+    /**
+     * @dev Variant of {parseInt-string} that parses a substring of `input` located between position `begin` (included) and
+     * `end` (excluded).
+     */
+    function parseInt(string memory input, uint256 begin, uint256 end) internal pure returns (int256) {
+        (bool success, int256 value) = tryParseInt(input, begin, end);
+        if (!success) revert StringsInvalidChar();
+        return value;
+    }
+
+    /**
+     * @dev Variant of {parseInt-string} that returns false if the parsing fails because of an invalid character.
+     *
+     * This function will still revert if the result does not fit in a `int256`
+     */
+    function tryParseInt(string memory input) internal pure returns (bool success, int256 value) {
+        return tryParseInt(input, 0, bytes(input).length);
+    }
+
+    /**
+     * @dev Variant of {parseInt-string-uint256-uint256} that returns false if the parsing fails because of an invalid
+     * character.
+     *
+     * This function will still revert if the result does not fit in a `int256`
+     */
+    function tryParseInt(
+        string memory input,
+        uint256 begin,
+        uint256 end
+    ) internal pure returns (bool success, int256 value) {
+        bytes memory buffer = bytes(input);
+
+        // check presence of a negative sign.
+        bool isNegative = bytes1(unsafeReadBytesOffset(buffer, begin)) == 0x2d;
+        int8 factor = isNegative ? int8(-1) : int8(1);
+        uint256 offset = isNegative.toUint();
+
+        int256 result = 0;
+        for (uint256 i = begin + offset; i < end; ++i) {
+            uint8 chr = _tryParseChr(buffer[i]);
+            if (chr > 9) return (false, 0);
+            result *= 10;
+            result += factor * int8(chr);
```

Comment:
I guess we can apply the factory until the end, but I know that might increase the scope right now since we'd need to check for the limits `[-(2**255), 2**255)`.

It's ok rn but we can consider for an optimization

Reply:
`int256` range from
`-2**255  = -57896044618658097711785492504343953926634992332820282019728792003956564819968`
to
`2**255-1 = 57896044618658097711785492504343953926634992332820282019728792003956564819967`

This means MinUint256 would not fit in a int256 if we don't consider the sign from the start. An alternative would be to parse as a `uint256` ... but then I'm not sure about how to do the casting

Reply:
I'd do something like

```solidity
    uint256 private constant ABS_MIN_INT256 = 2 ** 255;

    function tryParseInt(
        string memory input,
        uint256 begin,
        uint256 end
    ) internal pure returns (bool success, int256 value) {
        bytes memory buffer = bytes(input);

        // check presence of a negative sign.
        bool positiveSign = bytes1(unsafeReadBytesOffset(buffer, begin)) == bytes1("+");
        bool negativeSign = bytes1(unsafeReadBytesOffset(buffer, begin)) == bytes1("-");
        uint256 offset = (positiveSign || negativeSign).toUint();

        (bool subSuccess, uint256 absValue) = tryParseUint(input, begin + offset, end);

        if (subSuccess && !negativeSign && absValue < ABS_MIN_INT256) {
            return (true, int256(absValue));
        } else if (subSuccess && negativeSign && absValue <= ABS_MIN_INT256) {
            return (true, int256(-absValue));
        } else {
            return (false, 0);
        }
    }
```
but the negative uint256 -> int256 case is not clear to me.

Reply:
Maybe

```
if (subSuccess && negativeSign && absValue == ABS_MIN_INT256) return (true, type(int256).min);
else if (subSuccess && absValue < ABS_MIN_INT256) return (true, negativeSign ? -int256(absValue) : int256(absValue));
else return (false, 0);